File: [23.12]Design-Pattern/Java/src/structural/decorator/Main.java
package structural.decorator;

public class Main {
    public static void main(String[] args) {
        Hamburger hamburger = new BasicHamBurger();
        System.out.println(hamburger.getName());
        System.out.println();

        hamburger = new CheeseDecorator(hamburger);
        System.out.println("치즈 토핑 추가");
        System.out.println(hamburger.getName());
        System.out.println();

        hamburger = new BulgogiDecorator(hamburger);
        System.out.println("불고기 토핑 추가");
        System.out.println(hamburger.getName());
        System.out.println();

        hamburger = new BulgogiDecorator(hamburger);
        System.out.println("불고기 토핑 추가");
        System.out.println(hamburger.getName());
        System.out.println();

        hamburger = new CheeseDecorator(hamburger);
        System.out.println("치즈 토핑 추가");
        System.out.println(hamburger.getName());

    }
}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/structural/decorator/Main.kt
package structural.decorator

fun main() {
    var hamburger: Hamburger = BasicHamburger()
    println(hamburger.getName())
    
    println()
    println("치즈 추가")
    hamburger = CheeseDecorator(hamburger)
    println(hamburger.getName())
    
    println()
    println("불고기 추가")
    hamburger = BulgogiDecorator(hamburger)
    println(hamburger.getName())
}
